draft-ietf-privacypass-batched-tokens.md

```diff
@@ -319,13 +327,116 @@ struct {
 If the FinalizeBatch function fails, the Client aborts the protocol. Token
 verification works exactly as specified in {{ISSUANCE}}.
 
+# Arbitrary Batched Verifiable Token
+
+## Client-to-Issuer Request {#arbitrary-client-to-issuer-request}
+
+The Client first creates all TokenRequest it wants to batch. To do so, the client follows
+protocol describing issuance, such as {{ISSUANCE, Section 5.1}} or {{ISSUANCE, Section 6.1}}.
+
+The Client then creates a BatchedTokenRequest structured as follows:
+
+~~~tls
+struct {
+  uint16_t token_type;
+  TokenRequest token_requests<0..2^16-1>;
+} BatchedTokenRequest
+~~~
+
+The structure fields are defined as follows:
+
+- "token_type" is a 2-octet integer, which matches the type of the TokenRequests in the batch.
+
+- "token_requests" are serialized TokenRequests, in network byte order. The number of token_requests, as a 2-octet integer, is prepended to the serialized TokenRequests. In addition, the 2-octet integer length of each TokenRequest is prepended to the serialized TokenRequests.
+
+The Client MUST ensure all token_requests have the same token_type, which is the same as the BatchedTokenRequest.token_type.
+
+The Client then generates an HTTP POST request to send to the Issuer Request
+URL, with the BatchTokenRequest as the content. The media type for this request
+is "application/private-token-batch-request". An example request for the Issuer
+Request URL "https://issuer.example.net/request" is shown below.
+
+~~~
+POST /request HTTP/1.1
+Host: issuer.example.net
+Accept: application/private-token-batch-response
+Content-Type: application/private-token-batch-request
+Content-Length: <Length of BatchTokenRequest>
+
+<Bytes containing the BatchTokenRequest>
+~~~
+
+## Issuer-to-Client Response {#arbitrary-issuer-to-client-response}
+
+Upon receipt of the request, the Issuer validates the following conditions:
+
+- The BatchTokenRequest contains a supported token_type registered with IANA.
+- The BatchTokenRequest contains token_requests with the same token_type, matching BatchTokenRequest.token_type.
+
+If any of these conditions is not met, the Issuer MUST return an HTTP 400 error
+to the client.
+For performance, an Issuer MAY return an HTTP 400 error if all token_requests are
+not targetted at the same cryptographic material.
+
+The Issuer then tries to deserialize the i-th element
+of BatchTokenRequest.token_requests using the protocol associated to BatchTokenRequest.token_type.
+If this fails for any of
+the BatchTokenRequest.token_requests values, the Issuer MUST return an HTTP 400
+error to the client. Otherwise, if the Issuer is willing to produce a token to
+the Client, the issuer creates a BatchTokenResponse structured as follows:
+
+
+~~~tls
+struct {
+  TokenResponse token_responses<0..2^16-1>;
```

Comment:
The proposal here is to fail.

If we'd like to support multiple requests, some succeeding, some failing, having the client provide an index for each request seems like a good solution.
This can be added to `BlindedElement`

Reply:
I've updated the data structure so that I define a `TokenRequest` with `token_type` and an opaque element

Reply:
for a first pass, this has been updated as follow:

issuer reads token_requests as they arrive. if the i-th fails or is not supports, they return a zero-sized element.

in addition, token_request have token_type as their first element. this is enforced by the issuer.